ddl/table.go

```diff
@@ -152,6 +138,138 @@ func onCreateTable(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error)
 	}
 }
 
+func inheritPlacementRuleFromDB(tbInfo *model.TableInfo, dbInfo *model.DBInfo) error {
+	// Can not use both a placement policy and direct assignment. If you alter specify both in a CREATE TABLE or ALTER TABLE an error will be returned.
+	if tbInfo.DirectPlacementOpts != nil && tbInfo.PlacementPolicyRef != nil {
+		return ErrPlacementPolicyWithDirectOption.GenWithStackByArgs(tbInfo.PlacementPolicyRef.Name)
+	}
+	if tbInfo.DirectPlacementOpts == nil && tbInfo.PlacementPolicyRef == nil {
+		if dbInfo.DirectPlacementOpts != nil {
+			clone := *dbInfo.DirectPlacementOpts
+			tbInfo.DirectPlacementOpts = &clone
+		}
+		if dbInfo.PlacementPolicyRef != nil {
+			clone := *dbInfo.PlacementPolicyRef
+			tbInfo.PlacementPolicyRef = &clone
+		}
+	}
+	return nil
+}
+
+func getBundleFromTblInfo(t *meta.Meta, job *model.Job, tbInfo *model.TableInfo) (*placement.Bundle, error) {
+	tableBundleWrapper := func(bundle *placement.Bundle) *placement.Bundle {
+		bundle.Index = placement.RuleIndexTable
+		bundle.Override = true
+		return bundle
+	}
+	var (
+		bundle *placement.Bundle
+		err    error
+	)
+	if tbInfo.DirectPlacementOpts != nil {
+		// check the direct placement option compatibility.
+		if err = checkPolicyValidation(tbInfo.DirectPlacementOpts); err != nil {
+			job.State = model.JobStateCancelled
+			return nil, errors.Trace(err)
+		}
```

Comment:
I've changed `checkPolicyValidation` into `NewBundleFromOptions` for checking. There is no need for this function here.